### PR TITLE
[backend] Add Saml2 auth option and upgrade User / Principal across t…

### PR DESCRIPTION
**Pull request by guillaumejparis**
…he app to accommodate this new way of authenticate  (#300)

### CODE REVIEW

**Comment by guillaumejparis** [comment]
Will fix https://github.com/OpenEx-Platform/openex/issues/362 too

File: openex-model/pom.xml [hunk 1]
```diff
@@ -71,5 +71,24 @@
...
             <version>2.0.1.Final</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-saml2-service-provider</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.opensaml</groupId>
+            <artifactId>opensaml-core</artifactId>
+            <version>4.1.1</version>
```
Comment (RomuDeuxfois) [inline openex-model/pom.xml:81]: Not sure if we need this in model package

File: openex-api/src/test/java/io/openex/runner/InitAdminCommandLineRunnerTest.java [hunk 2]
```diff
@@ -20,6 +22,9 @@
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
 public class InitAdminCommandLineRunnerTest {
 
+  @MockBean
+  private RelyingPartyRegistrationRepository relyingPartyRegistrationRepository;
```
Comment (RomuDeuxfois) [inline openex-api/src/test/java/io/openex/runner/InitAdminCommandLineRunnerTest.java:26]: Why we need it here ?

File: openex-api/src/main/java/io/openex/service/UserService.java [hunk 3]
```diff
@@ -28,86 +31,101 @@
...
-        user.setGroups(assignableGroups);
-        // Save the user
-        User savedUser = userRepository.save(user);
-        createUserToken(savedUser);
-        return savedUser;
+    SecurityContext context = SecurityContextHolder.createEmptyContext();
+    context.setAuthentication(new PreAuthenticatedAuthenticationToken(new OpenexPrincipal() {
+      @Override
+      public String getId() {
+        return user.getId();
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/service/UserService.java:81]: It seems a bit strange to me to have methods based on objects outside the created class

File: openex-api/src/main/java/io/openex/service/MailingService.java [hunk 4]
```diff
@@ -19,53 +20,61 @@
...
-    public void sendEmail(String subject, String body, List<User> users) {
-        sendEmail(subject, body, users, Optional.empty());
+  public void sendEmail(String subject, String body, List<User> users, Optional<Exercise> exercise) {
+    EmailContent emailContent = new EmailContent();
+    emailContent.setSubject(subject);
+    emailContent.setBody(body);
+    Inject inject = new Inject();
+    inject.setContent(mapper.valueToTree(emailContent));
+    inject.setContract(EmailContract.EMAIL_DEFAULT);
+    inject.setUser(userRepository.findById(currentUser().getId()).orElseThrow());
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/service/MailingService.java:63]: Can you add the keyword `this.` when you are using class's properties ?
It's more readable.

File: openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java [hunk 5]
```diff
@@ -28,115 +29,139 @@
...
+  }
 
-    private PlatformSetting buildPlatformSetting(Map<String, Setting> dbSettings, SETTING_KEYS setting) {
-        Optional<Setting> platformName = ofNullable(dbSettings.get(setting.key()));
-        String settingValue = platformName.map(Setting::getValue).orElse(setting.defaultValue());
-        return new PlatformSetting(setting.key(), settingValue);
+  private List<OAuthProvider> buildSaml2Providers() {
+    try {
+      Saml2RelyingPartyProperties properties = context.getBean(Saml2RelyingPartyProperties.class);
+      System.out.println(properties);
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java:87]: Remove SYSO

File: openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java [hunk 6]
```diff
@@ -28,115 +29,139 @@
...
+            if (clientName == null) {
+              clientName = "Login with " + entry.getKey();
+            }
+            return new OAuthProvider(entry.getKey(), uri, clientName);
+          })
+          .toList();
+    } catch (Exception e) {
+      // No provider defined in the configuration
+      return new ArrayList<>();
     }
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java:103]: We can have one buildProviders with line 89 to 103
Rename the actuel buildProviders to buildOauth2Providers
And call  buildProviders method inside buildOauth2Providers and buildSaml2Providers

File: openex-api/src/main/java/io/openex/rest/exercise/ExerciseApi.java [hunk 7]
```diff
@@ -55,583 +54,598 @@
...
+
+  @PostMapping("/api/exercises/import")
+  @RolesAllowed(ROLE_ADMIN)
+  public void exerciseImport(@RequestPart("file") MultipartFile file) throws Exception {
+    importService.handleFileImport(file);
+  }
+
+  @GetMapping("/api/player/exercises/{exerciseId}")
+  public PublicExercise playerExercise(@PathVariable String exerciseId, @RequestParam Optional<String> userId) {
+    impersonateUser(userRepository, userId); // TODO Check Security
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/rest/exercise/ExerciseApi.java:646]: TODO ?

File: openex-api/pom.xml [hunk 8]
```diff
@@ -180,6 +180,25 @@
             <!--suppress MavenPackageUpdate -->
             <version>1.15</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-saml2-service-provider</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.opensaml</groupId>
```
Comment (RomuDeuxfois) [inline openex-api/pom.xml:188]: You can extract this in properties variables

File: openex-api/src/main/java/io/openex/config/OpenexPrincipal.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package io.openex.config;
+
+import org.springframework.security.core.GrantedAuthority;
+
+import java.util.Collection;
+
+public abstract class OpenexPrincipal {
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/config/OpenexPrincipal.java:7]: Can be an interface

File: openex-api/src/main/java/io/openex/config/AppSecurityConfig.java [hunk 10]
```diff
@@ -26,165 +29,398 @@
...
+    Saml2AuthenticatedPrincipal user = (Saml2AuthenticatedPrincipal) authentication.getPrincipal();
+    User loginUser = userSaml2Management(user);
+
+    List<SimpleGrantedAuthority> roles = new ArrayList<>();
+    roles.add(new SimpleGrantedAuthority(ROLE_USER));
+    if (loginUser.isAdmin()) {
+      roles.add(new SimpleGrantedAuthority(ROLE_ADMIN));
+    }
+
+    return new Saml2Authentication(new OpenExSaml2User() {
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/config/AppSecurityConfig.java:378]: OpenExSaml2User can be a class to move all this next line and be more readable
What do you think ?

File: openex-api/src/main/java/io/openex/config/AppSecurityConfig.java [hunk 11]
```diff
@@ -26,165 +29,398 @@
...
+        return loginUser.getFirstname() + " " + loginUser.getLastname();
+      }
+
+    };
+  }
+
+  public OAuth2User oAuth2UserManagement(OAuth2AccessToken accessToken, ClientRegistration clientRegistration,
+      OAuth2User user) {
+    User loginUser = userOauth2Management(accessToken, clientRegistration, user);
+    return new OpenExOAuth2User() {
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/config/AppSecurityConfig.java:329]: OpenExOAuth2User can be a class to move all this next line and be more readable
What do you think ?

File: openex-api/src/main/java/io/openex/config/AppSecurityConfig.java [hunk 12]
```diff
@@ -26,165 +29,398 @@
...
     }
+
+    Saml2Error authError = new Saml2Error("invalid_token", "User conversion fail");
+    throw new Saml2AuthenticationException(authError);
+  }
+
+  public OidcUser oidcUserManagement(OAuth2AccessToken accessToken, ClientRegistration clientRegistration,
+      OAuth2User user) {
+    User loginUser = userOauth2Management(accessToken, clientRegistration, user);
+    return new OpenExOidcUser() {
```
Comment (RomuDeuxfois) [inline openex-api/src/main/java/io/openex/config/AppSecurityConfig.java:273]: OpenExOidcUser can be a class to move all this next line and be more readable
What do you think ?

**Comment by RomuDeuxfois** [comment]
> Will fix #362 too

Test 🆗

File: openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java (hunk 5 above)
Comment (guillaumejparis) [reply openex-api/src/main/java/io/openex/rest/settings/SettingsApi.java:87]:
```suggestion
```

File: openex-api/src/main/java/io/openex/config/OpenexPrincipal.java (hunk 9 above)
Comment (guillaumejparis) [reply openex-api/src/main/java/io/openex/config/OpenexPrincipal.java:7]:
```suggestion
public interface OpenexPrincipal {
```

**Review by RomuDeuxfois — APPROVED** [review]

**Pull request merged**